WalletWasabi.Tests/UnitTests/WabiSabi/Models/MultipartyTransactionTests.cs

```diff
@@ -0,0 +1,327 @@
+
+			Assert.Single(differentInput.Inputs);
+			Assert.Empty(differentInput.Outputs);
+			Assert.NotEqual(oneInput.Inputs, differentInput.Inputs);
+			Assert.Equal(oneInput.Outputs, differentInput.Outputs);
+
+			var twoInputs = oneInput.AddInput(alice2Coin);
+
+			Assert.Equal(2, twoInputs.Inputs.Count);
+			Assert.Empty(twoInputs.Outputs);
+
+			// address reuse bad
+			var bob1 = new TxOut(Money.Coins(1), alice1Coin.ScriptPubKey);
+			var withOutput = twoInputs.AddOutput(bob1);
+
+			Assert.Equal(2, withOutput.Inputs.Count);
+			Assert.Single(withOutput.Outputs);
+
+			var bob2 = new TxOut(Money.Coins(1), alice2Coin.ScriptPubKey);
+			var noFeeTx = withOutput.AddOutput(bob2).Finalize();
+			Assert.Equal(Money.Zero, noFeeTx.Balance);
+
+			var tx = noFeeTx.CreateUnsignedTransaction();
+			Assert.Equal(2, tx.Inputs.Count);
+			Assert.Equal(2, tx.Outputs.Count);
+			Assert.Contains(alice1Coin.Outpoint, tx.Inputs.Select(x => x.PrevOut));
+			Assert.Contains(alice2Coin.Outpoint, tx.Inputs.Select(x => x.PrevOut));
+			Assert.Contains(alice1Coin.ScriptPubKey, tx.Outputs.Select(x => x.ScriptPubKey));
+			Assert.Contains(alice2Coin.ScriptPubKey, tx.Outputs.Select(x => x.ScriptPubKey));
+
+			var alice1Tx = tx.Clone();
+			alice1Tx.Sign(key1.GetBitcoinSecret(Network.Main), alice1Coin);
+
+			var alice1Sig = noFeeTx.AddWitness(0, alice1Tx.Inputs[0].WitScript);
+			Assert.False(alice1Sig.IsFullySigned);
+			Assert.Equal(alice1Tx.ToString(), alice1Sig.CreateTransaction().ToString());
+
+			var alice2Tx = tx.Clone();
+			alice2Tx.Sign(key2.GetBitcoinSecret(Network.Main), alice2Coin);
+
+			var alice2Sig = alice1Sig.AddWitness(1, alice2Tx.Inputs[1].WitScript);
+			Assert.True(alice2Sig.IsFullySigned);
+
+			var signed = alice2Sig.CreateTransaction();
+			Assert.NotEqual(alice1Tx.ToString(), signed.ToString());
+			Assert.NotEqual(alice2Tx.ToString(), signed.ToString());
+			Assert.True(signed.Inputs.All(x => x.HasWitScript()));
+		}
+
+		[Fact]
+		public void AddWithOptimize()
+		{
+			var alice = WabiSabiFactory.CreateAlice();
+
+			var state = new Construction(DefaultParameters).AddInput(alice.Coins.First());
+
+			var script = BitcoinFactory.CreateScript();
+			var bob = new TxOut(alice.Coins.First().Amount/2, script);
+			var withOutput = state.AddOutput(bob);
+			var duplicateOutputNoFee = withOutput.AddOutput(bob).Finalize();
+
+			var tx2 = duplicateOutputNoFee.CreateUnsignedTransaction();
+			var output = Assert.Single(tx2.Outputs);
+			Assert.Equal(script, output.ScriptPubKey);
+			Assert.Equal(alice.Coins.First().Amount, output.Value);
+		}
+
+		[Fact]
+		public void WitnessValidation()
+		{
+			using Key key1 = new();
+			using Key key2 = new();
+
+			var alice1 = WabiSabiFactory.CreateAlice(key: key1);
+			var alice2 = WabiSabiFactory.CreateAlice(key: key2);
+
+			var alice1Coin = alice1.Coins.First();
+			var alice2Coin = alice2.Coins.First();
+
+			var state = new Construction(DefaultParameters).AddInput(alice1Coin).AddInput(alice2Coin);
```

Comment:
Have you considered using the fluent builder pattern? https://medium.com/@jacobcunningham/the-fluent-builder-pattern-ac1b6c23afc3

with static?

Reply:
Thanks for sharing I was not familiar with it. If I understand correctly this is somewhat similar in its goal (getting support from the type system for ensuring a state machine is followed in the correct sequence).

FWIW this set of classes is not exactly a builder, although it does build transactions, its main purpose is to constrain those transactions to be standard and fair in a multiparty setting (so it's the intermediate result of processing a list of ordered events). If we want to support more complicated scripts in the future this will become more complex (e.g. computing the transaction's version and nLocktime as a function of all the inputs to support bip 112 or bip 65, see bip 370 for examples of the kind of logic that might be required). This is also the place to enforce specific rules/conditions on the transaction in the round.

We need 2-phase state updates because we have multiple things going on in the arena, this is why this was written with immutable collections. The example in the article uses a mutable class, but I think we can still use the pattern of returning `this` under multiple constrained interfaces, but if you're making a sequence of immutable objects there's no need for interfaces since it's just a new statically typed object for every step. I initially tried this approach but it requires algebraic data types to do cleanly. If C# supported an enum of record types for example (i.e. sum of product types) we could do it in a completely statically typed way, but encode this sum type as two sibling record classes inheriting from an abstract base class that represents the sum is the closest approximation I could come up with.

However, regardless of which approach to leverage the type system, there will still be the problem that the mutable `Round` property needs to be of the "sum" type, i.e. it must be able to contain any intermediate state, so we can't do just static types throughout (hence the Assert methods). Perhaps we could consider applying the same approach to `Round` itself, and have different interfaces for each round phase, `Arena` could then keep different rounds in different collections based on their type (and no need for a `Phase` enum), and only certain actions would be possible on the round enforced by the type system. I think that would be a cleaner approach than having a single shared round object for all phases, and it would allow using the same trick that a sequence of interface types statically determines the valid sequence for all sub-objects, both the backend specific properties and this shared multiparty transaction model. Unfortunately this would be a very big departure from the way the code is currently laid out, and even if it is a good idea (i'm not sure) I don't think now is the right time, it's a lot of work. Perhaps we can revisit after a minimal end to end test?

Reply:
This is a new concept in the project, so it has to be discussed in advance. Don't forget about the maintenance and teaching new developers about our actual code in the future. This is not something easy to understand, neither look familiar with any patterns the way it is now - at least for me. 

However, the clear separation of states and actions is something I would encourage. I think it is possible to make it more digestible, like with Lucas's suggestion. 

What do you think about having a base class instead of the interface to remove duplicates like: 

```
		public Money Balance => Inputs.Sum(x => x.Amount) - Outputs.Sum(x => x.Value);
		public int EstimatedInputsVsize => Inputs.Sum(x => x.TxOut.ScriptPubKey.EstimateInputVsize());
		public int OutputsVsize => Outputs.Sum(x => x.ScriptPubKey.EstimateOutputVsize());

		public int EstimatedVsize => MultipartyTransactionParameters.SharedOverhead + EstimatedInputsVsize + OutputsVsize;
		public FeeRate EffectiveFeeRate => new FeeRate(Balance, EstimatedVsize);
```

Reply:
I tried that both with state as an interface and as an abstract base class. I thought it it would work with the latter at least types but apparently that does't cover properties, and I don't quite understand why default impementations of properties didn't work (need to reread the spec on that) because I thought even with an interface it'd work.

The underlying issue is that the `Inputs` and `Outputs` properties are of different types (`ImmutableList`, which can be appended to in `ConstructionState` vs. `ImmutableArray`, which is fixed). I went with keeping these types different at the cost of some code duplication because it seems more correct to me. I could experiment more, for example using straight methods instead of properties should probably allow the base class to specify the collections as `IEnumerable<Coin>`, or just making them the same type would allow this deduplication.

Reply:
Coding conventions: try to avoid implementations in interfaces, hard to read the code - instead use base classes or extension methods.

Reply:
Alright, I missed that the types are different. Then it is fine from my side.